### ISSUE TITLE
3.0: nvliblist.conf hardcoded nvidia binary path

### ISSUE THREAD

**Issue by cclerget** [apptainer/singularity#2133]
## Version of Singularity:

3.0

### Expected behavior

Install nvidia binaries in non standard PATH and bind them with `--nv`

### Actual behavior

`nvliblist.conf` have hardcoded binaries path and search for exact `/usr/bin/<nvbinary>`, so if binaries are installed in a non standard PATH, they can't be bound because of hardcoded path.

### Steps to reproduce behavior

Install nvidia binaries in non standard PATH like `/usr/local/bin`

**Comment by jmstover** [comment]
I see this as ... expected behavior. nvliblist.conf is an admin controlled file, so they put in what libraries/binaries to include.

We filled it out with a system install, but if that isn't how the install was done...

**Comment by GodloveD** [comment]
Yes I agree with @jmstover.  This file is editable by the admin to point at whatever binaries they want.  Also, this file is not even used if nvidia-containter-cli is available on the system.  The nvidia-container-cli package is supposed to be shipped with nvidia drivers in the future.  So as it becomes more widely available, the conf file should become obsolete.

**Comment by cclerget** [comment]
@jmstover @GodloveD But this is not a consistent behaviour, if we consider a default installation why just put libraries name in `nvliblist.conf` rather than full path (and retrieve full path with ldconfig) ? and on the other side why admin should provide full path to nvidia binaries ?

this is more consistent and flexible to use `LookPath` to search binaries (like it's already done to retrieve path to `nvidia-containter-cli` in `nvidiautils.go`), so admins doesn't need to provide full path for binaries and end up with a file like :

```
nvidia-smi
nvidia-debugdump
nvidia-persistenced
nvidia-cuda-mps-control
nvidia-cuda-mps-server

# put libs here (must end in .so) 
libcuda.so
libEGL_installertest.so
libEGL_nvidia.so
```

rather than :
``````
/usr/bin/nvidia-smi
/usr/bin/nvidia-debugdump
/usr/bin/nvidia-persistenced
/usr/bin/nvidia-cuda-mps-control
/usr/bin/nvidia-cuda-mps-server

# put libs here (must end in .so) 
libcuda.so
libEGL_installertest.so
libEGL_nvidia.so
```

**Comment by jmstover** [comment]
You could have multiple versions of the libraries (that _could_ be in ldconfig cache), or not in ldconfig at all just controlled with `LD_LIBRARY_PATH`.

> to use LookPath

This is only going to search _system_ paths. Nothing outside that. So, _if_ the nvidia driver is installed under say:  /global/utils/nvida/[ver]  ... they won't be found.

**Comment by cclerget** [comment]
@jmstover According to documentation :

> LookPath searches for an executable named file in the directories named by the PATH environment variable

Otherwise even `nvidia-containter-cli` couldn't be found for non standard installation ...

**Comment by jmstover** [comment]
Exactly ... and we are sanitizing the PATH on run to only include:
`/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/sbin`

Just as 2.x does. You're saying "Just use LookPath" ... but the PATH where installed may not be in the PATH when ran.

`nvliblist.conf` is the _fallback_ (nvidia-container-cli is primary)... and should be the full path of binaries/libraries the admin wants to include if `nvidia-container-cli` fails.

**Comment by ctmadison** [comment]
This works the same as 2.x does. By design.

**Comment by GodloveD** [comment]
@cclerget what about this idea?  

We allow the user _either_ to specify a full path _or_ to just give the name of a file in `nvliblist.conf`.  We do this for both libraries and binaries. If the conf file entry contains a `/`, then we just trust that the user has provided the appropriate path and try to bind that.  If the entry does not contain a `/` we search for the file in ldconfig output (library) or using `LookPath` (binary).  So the admin can simplify the conf file if they have a standard installation, but they can also provide the full path if they have a non-standard installation.  

How does that sound?

**Comment by cclerget** [comment]
@GodloveD In fact `LookPath` can be called whatever if it's an absolute or just filename, according to `LookPath` documentation `If file contains a slash, it is tried directly and the PATH is not consulted`, so we can always call `LookPath` for binaries and avoid useless checks. I finally included changes in #2156 to maximize chance to find `nvidia-containter-cli` and other binaries with non standard installation. So before reset `PATH` in singularity CLI, just backup user `PATH` to `USER_PATH` and pass `USER_PATH` as argument to `GetNvidiaPath` which override current `PATH` and restore it when return, like that all `LookPath` works with the user defined paths that may be affected by module environment or ...

**Comment by ArangoGutierrez** [comment]
https://github.com/sylabs/singularity/pull/2156
